pkg/component/base/execution.go

```diff
@@ -162,94 +184,173 @@ func optionMatch(valueMap *structpb.Struct, reference map[string]interface{}, ch
-				var requiredFieldsRaw []interface{}
-				if requiredFieldsRaw, ok = valueMap["required"].([]interface{}); !ok {
-					continue
-				}
-				requiredFields := make([]string, len(requiredFieldsRaw))
-				for idx, v := range requiredFieldsRaw {
-					requiredFields[idx] = fmt.Sprintf("%v", v)
-				}
-				for _, target := range []string{"allOf", "anyOf", "oneOf"} {
-					var items []interface{}
-					if items, ok = valueMap[target].([]interface{}); !ok {
-						continue
-					}
-					for _, v := range items {
-						if properties, ok = v.(map[string]interface{})["properties"].(map[string]interface{}); !ok {
-							continue
-						}
-						inputSubField := input.GetFields()[key].GetStructValue()
-						if target == "oneOf" && !optionMatch(inputSubField, properties, requiredFields) {
-							continue
-						}
-						subField, err := e.fillInDefaultValuesWithReference(inputSubField, properties)
-						if err != nil {
-							return nil, err
-						}
-						input.GetFields()[key] = structpb.NewStructValue(subField)
-					}
-				}
-			} else {
-				if properties, ok = valueMap["properties"].(map[string]interface{}); !ok {
-					continue
-				}
-				subField, err := e.fillInDefaultValuesWithReference(input.GetFields()[key].GetStructValue(), properties)
-				if err != nil {
-					return nil, err
-				}
-				input.GetFields()[key] = structpb.NewStructValue(subField)
+
+			if err := handleProperties(e, inputStruct, key, properties); err != nil {
+				return err
 			}
-			continue
 		}
-		if _, ok := input.GetFields()[key]; ok {
+	}
+	return nil
+}
+
+func handleDefaultValue(inputStruct *structpb.Struct, key string, valueMap map[string]interface{}) error {
+	if _, exists := inputStruct.GetFields()[key]; exists {
+		return nil
+	}
+
+	defaultValue := valueMap["default"]
+	typeValue := valueMap["type"].(string)
+
+	switch typeValue {
+	case "string", "integer", "number", "boolean":
+		val, err := structpb.NewValue(defaultValue)
+		if err == nil {
+			inputStruct.GetFields()[key] = val
+		}
+	case "array":
+		if err := handleDefaultArray(inputStruct, key, valueMap); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func handleDefaultArray(inputStruct *structpb.Struct, key string, valueMap map[string]interface{}) error {
+	tempArray := &structpb.ListValue{Values: []*structpb.Value{}}
+	itemType := valueMap["items"].(map[string]interface{})["type"].(string)
+
+	if itemType != "string" && itemType != "integer" && itemType != "number" && itemType != "boolean" {
+		return nil
+	}
+
+	for _, v := range valueMap["default"].([]interface{}) {
+		if val, err := structpb.NewValue(v); err == nil {
+			tempArray.Values = append(tempArray.Values, val)
+		}
+	}
+
+	inputStruct.GetFields()[key] = structpb.NewListValue(tempArray)
+	return nil
+}
+
+// Main function
+func (e *ComponentExecution) fillInDefaultValuesWithReference(input any, reference map[string]interface{}) (any, error) {
+	inputStruct, err := convertToStructPb(input)
```

Comment:
nit: if this is only going to take and return a `structpb.Struct`, does it make sense to have `any` in the signature instead of that type?

Reply:
This is a bit tricky because we need to pass input `any` from [here](https://github.com/instill-ai/pipeline-backend/pull/872/files#diff-af6972698e144d60b8b3202a679de0954cca17bd5d92a3d3bc1f7a5debd93c21R138). If we convert input from `any` to `*structpb.Struct` there, we will need to unmarshal [this](https://github.com/instill-ai/pipeline-backend/pull/872/files#diff-27f7ebfb46d49db462803ce492d42f1af3cfebbc3efc7956a302d55a7ecff01dR346-R348) in the same `Execute` function.

I think we'd better refactor how default values are injected to the input variable given the new I/O interface. IMHO, the current way is not elegant. cc'd @donch1989

Reply:
> One thing I noticed is that every call to FillInDefaultValues will ignore the return value and only check the error. If that's the (only) intendeded use, perhaps we should simplify the signature. But we can do it when we refactor this part.

Good point. That was for the old implementation using with the internal recursive private function `fillInDefaultValuesWithReference`. I will take it out and rebase the latest main before merging.

Reply:
The fillInDefaultValuesWithReference indeed not works well under the current structure. I'm thinking we can utilize the go tag with our unmarhal function to handle this.
For example
```golang
type Output struct{
  str1 *string `instill:"str1,default=VAL"`
}
```